File: leetcode_75/Graphs_DFS/cpp/1466.cpp
#include <vector>
#include <unordered_map>

using namespace std;

class Solution
{
public:
    int minReorder(int n, vector<vector<int>> &connections)
    {
        unordered_map<int, vector<int>> base_graph;
        unordered_map<int, vector<int>> oriented_graph;
        for (int i = 0; i < n - 1; ++i)
        {
            base_graph[connections[i][0]].push_back(connections[i][1]);
            oriented_graph[connections[i][0]].push_back(connections[i][1]);
            oriented_graph[connections[i][1]].push_back(connections[i][0]);
        }

        int root = 0;
        vector<bool> visited(n, false);
        visited[0] = true;
        this->dfs(root, oriented_graph, base_graph, visited);

        return this->count;
    }

private:
    int count = 0;
    void dfs(
        int node,
        unordered_map<int, vector<int>> &oriented_graph,
        unordered_map<int, vector<int>> &base_graph,
        vector<bool> &visited)
    {
        visited[node] = true;
        for (auto v : oriented_graph[node])
        {
            if (!visited[v])
            {
                if (find(base_graph[node].begin(), base_graph[node].end(), v) != base_graph[node].end())
                {
                    this->count++;
                }
                this->dfs(v, oriented_graph, base_graph, visited);
            }
        }
    }
};

File: leetcode_75/Graphs_DFS/python/1466.py
from typing import List


class Solution:

    __count = 0

    def dfs(self, node, oriented_graph, base_graph, visited):
        visited[node] = True
        for v in oriented_graph[node]:
            if not visited[v]:
                if v in base_graph[node]:
                    self.__count += 1
                self.dfs(v, oriented_graph, base_graph, visited)

    def minReorder(self, n: int, connections: List[List[int]]) -> int:
        base_graph = {i: set() for i in range(n)}
        oriented_graph = {i: set() for i in range(n)}

        for i in range(n - 1):
            base_graph[connections[i][0]].add(connections[i][1])
            oriented_graph[connections[i][0]].add(connections[i][1])
            oriented_graph[connections[i][1]].add(connections[i][0])

        root = 0
        visited = [False] * n
        visited[0] = True
        self.dfs(root, oriented_graph, base_graph, visited)

        return self.__count
